File: src/moo/tokens.py
"""Token-Definitionen und zweisprachiges Keyword-Mapping für moo."""

from enum import Enum, auto


class TokenType(Enum):
    # Literals
    NUMBER = auto()
    STRING = auto()
    IDENTIFIER = auto()
    BOOLEAN = auto()
    NONE = auto()

    # Keywords
    SET = auto()        # setze / set
    TO = auto()         # auf / to
    IF = auto()         # wenn / if
    ELSE = auto()       # sonst / else
    WHILE = auto()      # solange / while
    FOR = auto()        # für / for
    IN = auto()         # in / in
    FUNC = auto()       # funktion / func
    RETURN = auto()     # gib_zurück / return
    SHOW = auto()       # zeige / show
    AND = auto()        # und / and
    OR = auto()         # oder / or
    NOT = auto()        # nicht / not
    CLASS = auto()      # klasse / class
    NEW = auto()        # neu / new
    THIS = auto()       # selbst / this
    IMPORT = auto()     # importiere / import
    FROM = auto()       # aus / from
    EXPORT = auto()     # exportiere / export
    TRY = auto()        # versuche / try
    CATCH = auto()      # fange / catch
    THROW = auto()      # wirf / throw
    BREAK = auto()      # stopp / break
    CONTINUE = auto()   # weiter / continue
    CONST = auto()      # konstante / const
    MATCH = auto()      # prüfe / match
    CASE = auto()       # fall / case
    DEFAULT = auto()    # standard / default
    AS = auto()         # als / as
    DATA = auto()       # daten / data
    AT = auto()         # @
    ASYNC = auto()      # asynchron / async
    AWAIT = auto()      # warte / await

    # Operators
    RANGE = auto()      # ..
    PLUS = auto()
    MINUS = auto()
    MULTIPLY = auto()
    DIVIDE = auto()
    MODULO = auto()       # %
    POWER = auto()        # **
    ASSIGN = auto()       # =
    PLUS_ASSIGN = auto()  # +=
    MINUS_ASSIGN = auto() # -=
    EQUALS = auto()       # ==
    NOT_EQUALS = auto()   # !=
    LESS = auto()         # <
    GREATER = auto()      # >
    LESS_EQ = auto()      # <=
    GREATER_EQ = auto()   # >=
    DOT = auto()          # .
    OPTIONAL_CHAIN = auto()   # ?.
    NULLISH_COALESCE = auto() # ??
    PIPE = auto()             # |>
    SPREAD = auto()           # ...

    # Delimiters
    COLON = auto()
    COMMA = auto()
    LPAREN = auto()
    RPAREN = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    LBRACE = auto()       # {
    RBRACE = auto()       # }
    ARROW = auto()        # =>

    # Special
    NEWLINE = auto()
    INDENT = auto()
    DEDENT = auto()
    EOF = auto()


# Zweisprachiges Keyword-Mapping: beide Sprachen → gleicher TokenType
KEYWORDS: dict[str, TokenType] = {
    # Deutsch
    "setze": TokenType.SET,
    "auf": TokenType.TO,
    "wenn": TokenType.IF,
    "sonst": TokenType.ELSE,
    "solange": TokenType.WHILE,
    "für": TokenType.FOR,
    "in": TokenType.IN,
    "funktion": TokenType.FUNC,
    "gib_zurück": TokenType.RETURN,
    "zeige": TokenType.SHOW,
    "und": TokenType.AND,
    "oder": TokenType.OR,
    "nicht": TokenType.NOT,
    "wahr": TokenType.BOOLEAN,
    "falsch": TokenType.BOOLEAN,
    "nichts": TokenType.NONE,
    "klasse": TokenType.CLASS,
    "neu": TokenType.NEW,
    "selbst": TokenType.THIS,
    "importiere": TokenType.IMPORT,
    "aus": TokenType.FROM,
    "exportiere": TokenType.EXPORT,
    "versuche": TokenType.TRY,
    "fange": TokenType.CATCH,
    "wirf": TokenType.THROW,
    "stopp": TokenType.BREAK,
    "weiter": TokenType.CONTINUE,
    "konstante": TokenType.CONST,
    "prüfe": TokenType.MATCH,
    "fall": TokenType.CASE,
    "standard": TokenType.DEFAULT,
    "als": TokenType.AS,
    "daten": TokenType.DATA,
    "asynchron": TokenType.ASYNC,
    "warte": TokenType.AWAIT,

    # English
    "set": TokenType.SET,
    "to": TokenType.TO,
    "if": TokenType.IF,
    "else": TokenType.ELSE,
    "while": TokenType.WHILE,
    "for": TokenType.FOR,
    "func": TokenType.FUNC,
    "return": TokenType.RETURN,
    "show": TokenType.SHOW,
    "and": TokenType.AND,
    "or": TokenType.OR,
    "not": TokenType.NOT,
    "true": TokenType.BOOLEAN,
    "false": TokenType.BOOLEAN,
    "none": TokenType.NONE,
    "class": TokenType.CLASS,
    "new": TokenType.NEW,
    "this": TokenType.THIS,
    "import": TokenType.IMPORT,
    "from": TokenType.FROM,
    "export": TokenType.EXPORT,
    "try": TokenType.TRY,
    "catch": TokenType.CATCH,
    "throw": TokenType.THROW,
    "break": TokenType.BREAK,
    "continue": TokenType.CONTINUE,
    "const": TokenType.CONST,
    "match": TokenType.MATCH,
    "case": TokenType.CASE,
    "default": TokenType.DEFAULT,
    "as": TokenType.AS,
    "data": TokenType.DATA,
    "async": TokenType.ASYNC,
    "await": TokenType.AWAIT,
}

# Boolean-Werte Mapping
BOOLEAN_VALUES: dict[str, bool] = {
    "wahr": True,
    "falsch": False,
    "true": True,
    "false": False,
}


class Token:
    __slots__ = ("type", "value", "line", "column")

    def __init__(self, type: TokenType, value: object, line: int, column: int):
        self.type = type
        self.value = value
        self.line = line
        self.column = column

    def __repr__(self) -> str:
        return f"Token({self.type.name}, {self.value!r}, L{self.line}:{self.column})"

File: compiler/src/tokens.rs
/// Token-Definitionen und zweisprachiges Keyword-Mapping für moo.

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    // Literals
    Number(f64),
    String(std::string::String),
    Boolean(bool),
    None,
    Identifier(std::string::String),

    // Keywords
    Set,        // setze / set
    To,         // auf / to
    If,         // wenn / if
    Else,       // sonst / else
    While,      // solange / while
    For,        // für / for
    In,         // in
    Func,       // funktion / func
    Return,     // gib_zurück / return
    Show,       // zeige / show
    And,        // und / and
    Or,         // oder / or
    Not,        // nicht / not
    Class,      // klasse / class
    New,        // neu / new
    This,       // selbst / this
    Try,        // versuche / try
    Catch,      // fange / catch
    Throw,      // wirf / throw
    Break,      // stopp / break
    Continue,   // weiter / continue
    Const,      // konstante / const
    Match,      // prüfe / match
    Case,       // fall / case
    Default,    // standard / default
    Import,     // importiere / import
    From,       // aus / from
    Export,     // exportiere / export
    As,         // als / as
    Defer,      // aufräumen / defer
    Data,       // daten / data
    Guard,      // garantiere / guard
    Where,      // wo / where
    Select,     // wähle / select
    Order,      // sortiere / order
    QueryFrom,  // von (Query-Kontext, nicht Import)
    Interface,  // schnittstelle / interface
    Implements, // implementiert / implements
    Parallel,       // parallel
    Precondition,   // vorbedingung / precondition
    Postcondition,  // nachbedingung / postcondition
    Unsafe,         // unsicher / unsafe
    Test,           // teste / test
    Expect,         // erwarte / expect

    // Bitwise Operators
    BitAnd,       // &
    BitOr,        // |
    BitNot,       // ~
    BitXor,       // ^
    LShift,       // <<
    RShift,       // >>

    // Operators
    At,           // @
    Plus,
    Minus,
    Multiply,
    Divide,
    Modulo,
    Range,        // ..
    Power,        // **
    Assign,       // =
    PlusAssign,   // +=
    MinusAssign,  // -=
    Equals,       // ==
    NotEquals,    // !=
    Less,
    Greater,
    LessEq,
    GreaterEq,
    Dot,
    OptionalChain,   // ?.
    NullishCoalesce, // ??
    Spread,          // ...
    Pipe,            // |>
    Question,        // ?

    // Delimiters
    Colon,
    Comma,
    LParen,
    RParen,
    LBracket,
    RBracket,
    LBrace,
    RBrace,
    Arrow,        // =>
    ThinArrow,    // -> (Rueckgabetyp)

    // Special
    Newline,
    Indent,
    Dedent,
    Eof,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub line: usize,
    pub column: usize,
}

impl Token {
    pub fn new(token_type: TokenType, line: usize, column: usize) -> Self {
        Self { token_type, line, column }
    }
}

/// Zweisprachiges Keyword-Mapping
pub fn keyword_lookup(word: &str) -> Option<TokenType> {
    match word {
        // Deutsch
        "setze" => Some(TokenType::Set),
        "auf" => Some(TokenType::To),
        "wenn" => Some(TokenType::If),
        "sonst" => Some(TokenType::Else),
        "solange" => Some(TokenType::While),
        "für" | "fuer" => Some(TokenType::For),
        "in" => Some(TokenType::In),
        "funktion" => Some(TokenType::Func),
        "gib_zurück" => Some(TokenType::Return),
        "zeige" => Some(TokenType::Show),
        "und" => Some(TokenType::And),
        "oder" => Some(TokenType::Or),
        "nicht" => Some(TokenType::Not),
        "wahr" => Some(TokenType::Boolean(true)),
        "falsch" => Some(TokenType::Boolean(false)),
        "nichts" => Some(TokenType::None),
        "klasse" => Some(TokenType::Class),
        "neu" => Some(TokenType::New),
        "selbst" => Some(TokenType::This),
        "versuche" => Some(TokenType::Try),
        "fange" => Some(TokenType::Catch),
        "wirf" => Some(TokenType::Throw),
        "stopp" => Some(TokenType::Break),
        "weiter" => Some(TokenType::Continue),
        "konstante" => Some(TokenType::Const),
        "prüfe" => Some(TokenType::Match),
        "fall" => Some(TokenType::Case),
        "standard" => Some(TokenType::Default),
        "importiere" => Some(TokenType::Import),
        "aus" => Some(TokenType::From),
        "exportiere" => Some(TokenType::Export),
        "als" => Some(TokenType::As),
        "aufräumen" => Some(TokenType::Defer),
        "daten" => Some(TokenType::Data),
        "garantiere" => Some(TokenType::Guard),
        "schnittstelle" => Some(TokenType::Interface),
        "implementiert" => Some(TokenType::Implements),
        "unsicher" => Some(TokenType::Unsafe),

        // Lern-Modus: ausfuehrliche Keywords fuer Anfaenger
        "setze_variable" => Some(TokenType::Set),
        "zeige_auf_bildschirm" => Some(TokenType::Show),
        "wenn_bedingung" => Some(TokenType::If),
        "sonst_alternative" => Some(TokenType::Else),
        "solange_wiederhole" => Some(TokenType::While),
        "fuer_jedes" => Some(TokenType::For),
        "funktion_definiere" => Some(TokenType::Func),
        "gib_wert_zurück" => Some(TokenType::Return),
        "neue_klasse" => Some(TokenType::Class),
        "importiere_modul" => Some(TokenType::Import),
        "versuche_ausfuehrung" => Some(TokenType::Try),
        "fange_fehler" => Some(TokenType::Catch),
        "teste" => Some(TokenType::Test),
        "erwarte" => Some(TokenType::Expect),
        "parallel" => Some(TokenType::Parallel),
        "vorbedingung" => Some(TokenType::Precondition),
        "nachbedingung" => Some(TokenType::Postcondition),
        "von" => Some(TokenType::QueryFrom),
        "wo" => Some(TokenType::Where),
        "wähle" => Some(TokenType::Select),
        "sortiere" => Some(TokenType::Order),

        // Experten-Kurzformen (2 Buchstaben)
        "se" => Some(TokenType::Set),
        "ze" => Some(TokenType::Show),
        "we" => Some(TokenType::If),
        "so" => Some(TokenType::Else),
        "sl" => Some(TokenType::While),
        "fu" => Some(TokenType::For),
        "fn" => Some(TokenType::Func),
        "kl" => Some(TokenType::Class),
        "gr" => Some(TokenType::Return),
        "ko" => Some(TokenType::Const),
        "st" => Some(TokenType::Break),
        "wt" => Some(TokenType::Continue),
        "im" => Some(TokenType::Import),
        "pr" => Some(TokenType::Match),
        "fa" => Some(TokenType::Case),
        "ve" => Some(TokenType::Try),
        // "fg" entfernt — kollidiert mit haeufigem Variablenname (foreground,
        // file-grep, frame-grab). 'fange' bzw. 'catch' ausgeschrieben verwenden.
        "wi" => Some(TokenType::Throw),
        "un" => Some(TokenType::Unsafe),
        // "wa" => Await existiert nicht im Rust-Compiler Token-Enum

        // English
        "set" => Some(TokenType::Set),
        "to" => Some(TokenType::To),
        "if" => Some(TokenType::If),
        "else" => Some(TokenType::Else),
        "while" => Some(TokenType::While),
        "for" => Some(TokenType::For),
        "func" => Some(TokenType::Func),
        "return" => Some(TokenType::Return),
        "show" => Some(TokenType::Show),
        "and" => Some(TokenType::And),
        "or" => Some(TokenType::Or),
        "not" => Some(TokenType::Not),
        "true" => Some(TokenType::Boolean(true)),
        "false" => Some(TokenType::Boolean(false)),
        "none" => Some(TokenType::None),
        "class" => Some(TokenType::Class),
        "new" => Some(TokenType::New),
        "this" => Some(TokenType::This),
        "try" => Some(TokenType::Try),
        "catch" => Some(TokenType::Catch),
        "throw" => Some(TokenType::Throw),
        "break" => Some(TokenType::Break),
        "continue" => Some(TokenType::Continue),
        "const" => Some(TokenType::Const),
        "match" => Some(TokenType::Match),
        "case" => Some(TokenType::Case),
        "default" => Some(TokenType::Default),
        "import" => Some(TokenType::Import),
        "from" => Some(TokenType::From),
        "export" => Some(TokenType::Export),
        "as" => Some(TokenType::As),
        "defer" => Some(TokenType::Defer),
        "data" => Some(TokenType::Data),
        "guard" => Some(TokenType::Guard),
        "interface" => Some(TokenType::Interface),
        "implements" => Some(TokenType::Implements),
        "unsafe" => Some(TokenType::Unsafe),
        "test" => Some(TokenType::Test),
        "expect" => Some(TokenType::Expect),
        "precondition" => Some(TokenType::Precondition),
        "postcondition" => Some(TokenType::Postcondition),
        "where" => Some(TokenType::Where),
        "select" => Some(TokenType::Select),
        "order" => Some(TokenType::Order),

        _ => Option::None,
    }
}
